### PR TITLE
[helm] Introduce glue for the less copy&pasta approach

### PR DESCRIPTION
**Pull request by lucendio**
* added Makefile and docs
* requires a `helmfile.yaml` in ${ENV_DIR}

TODO:
* [x] Add Helmfile to nix as a dependency

The first counter part can be found in https://github.com/zinfra/cailleach/pull/515

### CODE REVIEW

**Comment by lucendio** [comment]
I'd like to consolidate all three Makefiles ([Terraform](https://github.com/wireapp/wire-server-deploy/blob/develop/terraform/environment/Makefile), [Ansible](https://github.com/wireapp/wire-server-deploy/blob/develop/ansible/Makefile), Helm) into one and move it into the top-level. Any thoughts on this? /cc @akshaymankar @jschaul

**Comment by jschaul** [comment]
> I'd like to consolidate all three Makefiles ([Terraform](https://github.com/wireapp/wire-server-deploy/blob/develop/terraform/environment/Makefile), [Ansible](https://github.com/wireapp/wire-server-deploy/blob/develop/ansible/Makefile), Helm) into one and move it into the top-level. Any thoughts on this? /cc @akshaymankar @jschaul

No strong opinions here. Sounds fine to me to consolidate them.

**Comment by akshaymankar** [comment]
> I'd like to consolidate all three Makefiles ([Terraform](https://github.com/wireapp/wire-server-deploy/blob/develop/terraform/environment/Makefile), [Ansible](https://github.com/wireapp/wire-server-deploy/blob/develop/ansible/Makefile), Helm) into one and move it into the top-level. Any thoughts on this? /cc @akshaymankar @jschaul

Sounds like a good idea :100:

**Comment by lucendio** [comment]
Great. I'll do this as a follow-up. Anyway, ready to be reviewed now.

File: helm/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+## Deploy environment created by `terraform/environment`
+
+An 'environment' is supposed to represent all the setup required for the Wire
+platform to function.
+
+'Deploying' an environment means creating the respective Objects on top of Kubernetes
+to instantiate all the services that together represent the Wire backend.
+playbooks against servers specified in an inventory, resulting in a fully This action
+can be re-run as often as you want (e.g. in case you change some variables or upgrade
+to new versions).
```
Comment (jschaul) [inline helm/README.md:24]: There is some problem here with these sentences, looks like a copy paste snippet of text at the wrong place.

File: helm/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Helm-based deployment
+
+The Wire platform is deployed on top of Kubernetes. This certainly includes all *stateless* services (e.g. Brig, Galley),
+but may or may not include *stateful* backing services (e.g. Cassandra or Elasticsearch).
+
+The respective Kubernetes objects are defined in Helm charts. This allows to template and transfer them. 
+The charts themselves are defined in [wire-server](https://github.com/wireapp/wire-server/tree/master/charts)
+and uploaded two the [release](https://s3-eu-west-1.amazonaws.com/public.wire.com/charts) or
```
Comment (jschaul) [inline helm/README.md:8]:
```suggestion
and uploaded to the [release](https://s3-eu-west-1.amazonaws.com/public.wire.com/charts) or
```

**Review by jschaul — COMMENTED** [review]
Didn't try this out, as I'm not yet familiar with helm files (maybe someone else is and could review this?). Minor comments on the README.

**Comment by jmatsushita** [comment]
> I'd like to consolidate all three Makefiles (Terraform, Ansible, Helm) into one and move it into the top-level. Any thoughts on this? 
Definitely in favor of this!

Our current `kubespray` playbook outputs the kubeconfig in `inventory/$CLUSTER/artifacts/admin.conf` but this PR seems to expect `$(ENV_DIR)/kubeconfig` and `$(ENV_DIR)/kubeconfig.dec`. Can we make sure things align here?

**Comment by lucendio** [comment]
> Our current `kubespray` playbook outputs the kubeconfig in `inventory/$CLUSTER/artifacts/admin.conf` but this PR seems to expect `$(ENV_DIR)/kubeconfig` and `$(ENV_DIR)/kubeconfig.dec`. Can we make sure things align here?

While this is true for Kubespray, it can be changed by setting `artifacts_dir` and we already have at least [one](https://github.com/zinfra/cailleach/blob/master/environments/kube-ci/kubeconfig), that has it in its local top-level - facilitated by a [script](https://github.com/zinfra/cailleach/blob/master/environments/kube-ci/encrypt-kubeconfig.sh). With this PR, I just adapted the current reality.

I have another local branch, that transforms this script into a play added to `ansible/kubernetes.yaml`. But this only works when Sops + GnuPG works from within wire-server-deploy. I talked with @arianvp about the necessary steps. Long story short: I feel you and It'll happen along the road :)

File: helm/README.md (hunk 1 above)
Comment (lucendio) [reply helm/README.md]: Great catch. Fixed.

**Review by jschaul — APPROVED** [review]

**Pull request merged**